## Gather
    The data includes three tables from clincal trails for a new oral insulin

In [157]:
import pandas as pd

In [158]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

In [159]:
# make sure the tables below is not collapsed
#pd.set_option('display.max_rows',None)

## Assess
    This assessment includes visual assessment and programmatic assessment. 
### Visual Assessment
    Get an overview for the three tables

In [160]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country                                          contact  \
0    United States             951-719-9170ZoeWellish@superrito.com   
1    United States            PamelaSHill@cuvox.de+1 (217) 569-3204   
2    United States                 402-363-6804JaeMDebord@gustr.com   
3    United States       PhanBaLiem@jourrapide.com+1 (732) 636-8246   
4    United States                  334-515-7487TimNeudorf@cuvox.de   
..             ...                                              ...   
498  United States      207-477-0579MustafaLindstrom@jourrapide.com   
499  United States               928-284-4492RumanBisliev@gustr.com   
500  United States            816-223-6007JinkedeKeizer@teleworm.us   
501  United States  ChidaluOnyekaozulu@jourrapide.com1 360 443 2060   
502  United States             PatrickGersten@rhyta.com402-848-4923   

     birthdate  weight  height   bmi  
0    7/10/1976   121.7      66  19.6  
1     4/3/1967   118.8      66  19.2  
2    2/19/1980   177.8      71  24.8  
3    7/26/1951   220.9      70  31.7  
4    2/18/1928   192.3      27  26.1  
..         ...     ...     ...   ...  
498  4/10/1959   181.1      72  24.6  
499  3/26/1948   239.6      70  34.4  
500  1/13/1971   171.2      67  26.8  
501  2/13/1952   176.9      67  27.7  
502   5/3/1954   138.2      71  19.3  

[503 rows x 14 columns]

In [161]:
# calculate BMI for Tim Neudorf to see if the height of that person should be 72 not 27
weight_lbs = 192.3
height_in = 72
BMI = 703 * weight_lbs / (height_in * height_in)
BMI

26.077719907407406

In [162]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [163]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

### Programmatic Assessment


In [164]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [165]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [166]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [167]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [168]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [169]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [170]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Silva          2
Aranda         2
              ..
Nebeolisa      1
Maslov         1
Arsanukayev    1
Pérez          1
Laine          1
Name: surname, Length: 466, dtype: int64

In [171]:
patients.address.value_counts()

123 Main Street            6
2778 North Avenue          2
2476 Fulton Street         2
648 Old Dear Lane          2
3251 Radio Park Drive      1
                          ..
1990 Spring Avenue         1
945 Maple Avenue           1
1240 Summit Park Avenue    1
3300 Woodridge Lane        1
1557 Byers Lane            1
Name: address, Length: 483, dtype: int64

In [172]:
patients.query("surname == 'Doe'")

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state  zip_code        country                      contact birthdate  \
215    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
229    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
237    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
244    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
251    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   
277    NY   12345.0  United States  johndoe@email.com1234567890  1/1/1975   

     weight  height   bmi  
215   180.0      72  24.4  
229   180.0      72  24.4  
237   180.0      72  24.4  
244   180.0      72  24.4  
251   180.0      72  24.4  
277   180.0      72  24.4

In [173]:
patients.query("surname == 'Jakobsen'")

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
432      Houston        TX   77020.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438  11/25/1962   185.2      67   

      bmi  
24   24.4  
29   24.4  
432  29.0

In [174]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [175]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [176]:
sum(treatments.auralin.isnull())

0

In [177]:
sum(treatments.novodra.isnull())

0

### Quality

#### `patients` : 
    - zip code should be string data type not float, and some zip code only have 4 digits
    - The person Tim Neudorf's height should be 72 inches not 27 inches
    - full state name sometimes, and abbreviation other times
    - given name for patient id_9 should be David not Dsvid
    - address, city, state, country, contact and zip_code have 12 missing records
    - birthdate should be DateTime data type
    - index #210 patients weight is only 48.8 lbs
    - there are six records for patient named John Doe who live in the same address(duplicate records)
    - The person Jakob and Jake with same surname Jakobsen and same address(duplicate records)
    
#### `treatment` : 
    - missing hba1c_change variable
    - letter u in auralin and novodra table, and - in the missing cells
    - lower-case names do not match patient table
    - missing data, 280 records in treatment table vs 350 in patient table
    - data type for auralin and novodra should be int or float for easy calculation
    - inaccurate hba1c_change for patent name elliot richardson	(0.97 should be 0.47 maybe)
    
    
#### `adverse_reactions`:
    - lower-case names do not match patient table


### Tideness

#### `patients`:
    - contact column mix the email and phone number together
    
    
#### `treatment`:
    - auralin and novodra should put one column named treatment, and start and end dosage changes for different columns

#### `adverse_reactions`:
    - should merge with treatment to be one table
    

## Clean


In [178]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing data

#### `treatments` table: missing records 280 vs 350

#### *Define*: 
import the treatments_cut document and concatenate with treatments_cut table

In [179]:
# import treatments_cut.csv
treatments_cut = pd.read_csv('treatments_cut.csv')

In [180]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [181]:
treatments_cut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [182]:
# concatenate treatments_cut.csv to treatments_clean table to fill the missing records
treatments_clean = pd.concat([treatments_clean, treatments_cut], ignore_index = True)

#### Test

In [183]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  213 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


#### `treatments` table: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)¶

#### *Define*:
HbA1c changes = HbA1c_start - HbA1c_end

In [184]:
treatments_clean.hba1c_change = treatments_clean.hba1c_start - treatments_clean.hba1c_end

#### *Test*:


In [185]:
treatments_clean.hba1c_change.describe()

count    350.000000
mean       0.396286
std        0.059283
min        0.200000
25%        0.350000
50%        0.390000
75%        0.440000
max        0.540000
Name: hba1c_change, dtype: float64

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email. We will split the two variables to different columns, and drop the contact column. 

In [186]:
patients_clean['phone'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})',expand=True)

In [187]:
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+_]+@[a-zA-Z0-9_]+\.[a-zA-Z0-9_.][a-zA-Z]+)',expand=True)

In [188]:
patients_clean = patients_clean.drop(['contact'], axis = 1)

In [189]:
patients_clean

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi              phone  \
0    United States  7/10/1976   121.7      66  19.6       951-719-9170   
1    United States   4/3/1967   118.8      66  19.2  +1 (217) 569-3204   
2    United States  2/19/1980   177.8      71  24.8       402-363-6804   
3    United States  7/26/1951   220.9      70  31.7  +1 (732) 636-8246   
4    United States  2/18/1928   192.3      27  26.1       334-515-7487   
..             ...        ...     ...     ...   ...                ...   
498  United States  4/10/1959   181.1      72  24.6       207-477-0579   
499  United States  3/26/1948   239.6      70  34.4       928-284-4492   
500  United States  1/13/1971   171.2      67  26.8       816-223-6007   
501  United States  2/13/1952   176.9      67  27.7       360 443 2060   
502  United States   5/3/1954   138.2      71  19.3       402-848-4923   

                                 email  
0             ZoeWellish@superrito.com  
1                 PamelaSHill@cuvox.de  
2                 JaeMDebord@gustr.com  
3            PhanBaLiem@jourrapide.com  
4                  TimNeudorf@cuvox.de  
..                                 ...  
498    MustafaLindstrom@jourrapide.com  
499             RumanBisliev@gustr.com  
500          JinkedeKeizer@teleworm.us  
501  ChidaluOnyekaozulu@jourrapide.com  
502           PatrickGersten@rhyta.com  

[503 rows x 15 columns]

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

#### *Define*
Melt the auralin and novodra to treatment column and split the dose column to dose_start and dose_end

#### *Code*

In [190]:
treatments_clean = treatments_clean.melt(id_vars = ['given_name', 'surname', 'hba1c_start', 'hba1c_end','hba1c_change'], var_name = 'treatment',value_name = 'dose')


In [191]:
# delete the duplicate records when we melt two columns to one
treatments_clean = treatments_clean[treatments_clean.dose != '-']

In [192]:
treatments_clean

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0       veronika    jindrová         7.63       7.20          0.43   auralin   
3           skye  gormanston         7.97       7.62          0.35   auralin   
6         sophia      haugen         7.65       7.27          0.38   auralin   
7          eddie      archer         7.89       7.55          0.34   auralin   
9           asia     woźniak         7.76       7.37          0.39   auralin   
..           ...         ...          ...        ...           ...       ...   
688  christopher    woodward         7.51       7.06          0.45   novodra   
690        maret    sultygov         7.67       7.30          0.37   novodra   
694        lixue       hsueh         9.21       8.80          0.41   novodra   
696        jakob    jakobsen         7.96       7.51          0.45   novodra   
698        berta  napolitani         7.68       7.21          0.47   novodra   

          dose  
0    41u - 48u  
3    33u - 36u  
6    37u - 42u  
7    31u - 38u  
9    30u - 36u  
..         ...  
688  55u - 51u  
690  26u - 23u  
694  22u - 23u  
696  28u - 26u  
698  42u - 44u  

[350 rows x 7 columns]

In [193]:
# split dose column to two columns dose_start and dose_end
treatments_clean[['dose_start','dose_end']] = treatments_clean['dose'].str.split('-',1,expand = True)

In [194]:
treatments_clean = treatments_clean.drop('dose', axis = 1)

In [195]:
# remove 'u' in dose_start and dose_end, and convert the data type to int, so we can calculate the difference
treatments_clean['dose_start'] = treatments_clean['dose_start'].str[0:2]
treatments_clean['dose_end'] = treatments_clean['dose_end'].str[1:3]
treatments_clean['dose_start'] = treatments_clean['dose_start'].astype(str).astype(int)
treatments_clean['dose_end'] = treatments_clean['dose_end'].astype(str).astype(int)

In [196]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

   dose_start  dose_end  
0          41        48  
3          33        36  
6          37        42  
7          31        38  
9          30        36

#### *Test*

In [197]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 698
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   hba1c_start   350 non-null    float64
 3   hba1c_end     350 non-null    float64
 4   hba1c_change  350 non-null    float64
 5   treatment     350 non-null    object 
 6   dose_start    350 non-null    int64  
 7   dose_end      350 non-null    int64  
dtypes: float64(3), int64(2), object(3)
memory usage: 24.6+ KB


#### Adverse reaction should be part of the `treatments` table

#### *Define*
Merge the adverse_reaction column to `treatments` table, joining on given_name, surname

In [198]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on =['given_name', 'surname'], how = 'left')

In [199]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

   dose_start  dose_end adverse_reaction  
0          41        48              NaN  
1          33        36              NaN  
2          37        42              NaN  
3          31        38              NaN  
4          30        36              NaN

In [200]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   given_name        350 non-null    object 
 1   surname           350 non-null    object 
 2   hba1c_start       350 non-null    float64
 3   hba1c_end         350 non-null    float64
 4   hba1c_change      350 non-null    float64
 5   treatment         350 non-null    object 
 6   dose_start        350 non-null    int64  
 7   dose_end          350 non-null    int64  
 8   adverse_reaction  35 non-null     object 
dtypes: float64(3), int64(2), object(4)
memory usage: 27.3+ KB


####  Change given_name and surname in `treatments` table to patient_id in `patients` table 

#### *Define*
Isolate the patient ID and names in the patients table, then convert these names to lower case to join with treatments. 

In [201]:
id_names = patients[['patient_id','given_name','surname']]

In [202]:
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/generic.py:5303: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


In [203]:
treatments_clean = pd.merge(treatments_clean, id_names, on = ['given_name','surname'])

In [204]:
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

In [205]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment  dose_start  dose_end  \
0         7.63       7.20          0.43   auralin          41        48   
1         7.97       7.62          0.35   auralin          33        36   
2         7.65       7.27          0.38   auralin          37        42   
3         7.89       7.55          0.34   auralin          31        38   
4         7.76       7.37          0.39   auralin          30        36   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [206]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   dose_start        349 non-null    int64  
 5   dose_end          349 non-null    int64  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


### Quality

#### Zip code in `patients` table should be float not string, and it should be five digit

##### *Define*
Convert the float zip code type to string, remove.0 at the end of the number, and pad four digit zip code with a leading 0
##### *Code*

In [207]:
# change the float type to string and pad four digit zip code with a leading 0
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(width = 5, side = 'left', fillchar='0')

#### *Test*

In [208]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

#### *Define*
Replace the height of Tim Neudorf with 72 in

#### *Code*

In [209]:
patients_clean.height.replace(27,72, inplace = True)

##### Test

In [210]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  birthdate  weight  \
0  Rancho California  California    92390  United States  7/10/1976   121.7   
1          Armstrong    Illinois    61812  United States   4/3/1967   118.8   
2               York    Nebraska    68467  United States  2/19/1980   177.8   
3         Woodbridge          NJ    07095  United States  7/26/1951   220.9   
4             Dothan          AL    36303  United States  2/18/1928   192.3   

   height   bmi              phone                      email  
0      66  19.6       951-719-9170   ZoeWellish@superrito.com  
1      66  19.2  +1 (217) 569-3204       PamelaSHill@cuvox.de  
2      71  24.8       402-363-6804       JaeMDebord@gustr.com  
3      70  31.7  +1 (732) 636-8246  PhanBaLiem@jourrapide.com  
4      72  26.1       334-515-7487        TimNeudorf@cuvox.de

#### State names sometimes is full, sometimes is abbreviation

##### Define
change the state name of California, Illinois, Nebraska, Florida and New York to its abbrevation

##### Code

In [211]:
patients_clean.state.unique()

array(['California', 'Illinois', 'Nebraska', 'NJ', 'AL', 'Florida', 'NV',
       'CA', 'MO', 'New York', 'MI', 'TN', 'VA', 'OK', 'GA', 'MT', 'MA',
       'NY', 'NM', 'IL', 'LA', 'PA', 'CO', 'ME', 'WI', 'SD', 'MN', 'FL',
       'WY', 'OH', 'IA', 'NC', 'IN', 'CT', 'KY', 'DE', 'MD', 'AZ', 'TX',
       'NE', 'AK', 'ND', 'KS', 'MS', 'WA', 'SC', 'WV', 'RI', 'NH', 'OR',
       nan, 'VT', 'ID', 'DC', 'AR'], dtype=object)

In [212]:
patients_clean.state.replace(['California', 'Illinois', 'Nebraska','Florida', 'New York'], ['CA','IL','NE','FL','NY'], inplace = True)

##### Test

In [213]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
OK    13
LA    13
MI    13
NJ    12
VA    11
MS    10
WI    10
MN     9
AL     9
TN     9
IN     9
KY     8
WA     8
NC     8
MO     7
KS     6
NE     6
ID     6
NV     6
SC     5
IA     5
CT     5
AZ     4
CO     4
AR     4
RI     4
ME     4
ND     4
MD     3
WV     3
OR     3
DE     3
SD     3
VT     2
MT     2
DC     2
WY     1
AK     1
NH     1
NM     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define
Dsvid should be changed to David in `patient` table
##### Code

In [214]:
patients_clean.given_name.replace('Dsvid', 'David', inplace = True)

##### Test

In [215]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

          phone                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Wrong datatypes (assigned sex, state, zip_code, and birthdate columns) 

##### Define
Convert assigned sex and state to categorical data types. Convert birthdate to datetime data type. 

##### Code

In [216]:
# to category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# to date
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)


##### Test

In [217]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      503 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone         491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [218]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   dose_start        349 non-null    int64  
 5   dose_end          349 non-null    int64  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [219]:
patients_clean.phone = patients_clean.phone.str.replace(r'\D+', "").str.pad(11, fillchar = '1')

##### Test

In [220]:
patients_clean.phone.sample(10)

144    12605915755
130    18434940313
204    14052345350
138    18049081436
139    15133830516
407    14054705088
165    16627169586
93     17862340038
354    12015862848
398    17067774981
Name: phone, dtype: object

#### John Doe

##### Define
There are six same records under John Doe, we are going to remove last five just keep the first one
##### Code

In [221]:
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   
229         230         male       John     Doe  123 Main Street  New York   
237         238         male       John     Doe  123 Main Street  New York   
244         245         male       John     Doe  123 Main Street  New York   
251         252         male       John     Doe  123 Main Street  New York   
277         278         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   
229    NY    12345  United States 1975-01-01   180.0      72  24.4   
237    NY    12345  United States 1975-01-01   180.0      72  24.4   
244    NY    12345  United States 1975-01-01   180.0      72  24.4   
251    NY    12345  United States 1975-01-01   180.0      72  24.4   
277    NY    12345  United States 1975-01-01   180.0      72  24.4   

           phone              email  
215  11234567890  johndoe@email.com  
229  11234567890  johndoe@email.com  
237  11234567890  johndoe@email.com  
244  11234567890  johndoe@email.com  
251  11234567890  johndoe@email.com  
277  11234567890  johndoe@email.com

In [222]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      503 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone         491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [223]:
patients_clean = patients_clean.drop([229,237,244,251,277], axis = 0)

##### Test

In [224]:
patients_clean[patients_clean.surname == 'Doe']

patient_id assigned_sex given_name surname          address      city  \
215         216         male       John     Doe  123 Main Street  New York   

    state zip_code        country  birthdate  weight  height   bmi  \
215    NY    12345  United States 1975-01-01   180.0      72  24.4   

           phone              email  
215  11234567890  johndoe@email.com

#### Multiple records for Jakobsen, Gersten, Taylor
##### Define
Remove records for Jake Jakobsen, Pat Gersten, and Sandy Taylor, because those are short for Jacob, Patrick and Sandra.
##### Code

In [225]:
patients_clean.surname.value_counts()

Taylor         3
Jakobsen       3
Schiavone      2
Souza          2
Berg           2
              ..
Maslov         1
Arsanukayev    1
Pérez          1
Ribeiro        1
Laine          1
Name: surname, Length: 466, dtype: int64

In [226]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
282         283       female      Sandy  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
282    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

           phone                       email  
131  13044382648    SandraCTaylor@dayrep.com  
282  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

In [227]:
patients_clean = patients_clean.drop([282], axis = 0)

In [228]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
29   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi        phone                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
29   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [231]:
patients_clean = patients_clean.drop([29], axis = 0)

In [232]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97           98         male    Patrick  Gersten  2778 North Avenue  Burr   
502         503         male        Pat  Gersten  2778 North Avenue  Burr   

    state zip_code        country  birthdate  weight  height   bmi  \
97     NE    68324  United States 1954-05-03   138.2      71  19.3   
502    NE    68324  United States 1954-05-03   138.2      71  19.3   

           phone                     email  
97   14028484923  PatrickGersten@rhyta.com  
502  14028484923  PatrickGersten@rhyta.com

In [233]:
patients_clean = patients_clean.drop([502], axis = 0)

##### Test

In [238]:
patients_clean.surname.value_counts()

Taylor         2
Berg           2
Silva          2
Aranda         2
Ogochukwu      2
              ..
Maslov         1
Arsanukayev    1
Pérez          1
Ribeiro        1
Laine          1
Name: surname, Length: 466, dtype: int64

#### kgs instead of lbs for Zaitseva weight

##### Define
Covert Zaitseva's weight from kgs to lbs
##### Code

In [239]:
weight_kgs = patients_clean.weight.min()

In [242]:
weight_lbs = weight_kgs * 2.20462

In [243]:
patients_clean.replace({'weight': weight_kgs}, weight_lbs, inplace = True)

##### Test

In [245]:
patients_clean[patients_clean.surname == 'Zaitseva']


patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country  birthdate      weight  height  \
210  Wooster    OH    44691  United States 1938-11-26  107.585456      63   

      bmi        phone                          email  
210  19.1  13302022145  CamillaZaitseva@superrito.com

In [246]:
patients_clean.weight.min()

102.1